### PR TITLE
`ParallelExecution` step

### PR DESCRIPTION
**Pull request by gtopper**
[ML-7689](https://iguazio.atlassian.net/browse/ML-7689)

[ML-7689]: https://iguazio.atlassian.net/browse/ML-7689?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by ZeevRispler — APPROVED** [review]
LGTM

File: storey/flow.py [hunk 1]
```diff
@@ -1435,3 +1436,182 @@ def get_table(self, key):
...
+        if not runnables:
+            raise ValueError("ParallelExecution cannot be instantiated without at least one runnable")
+
+        self.runnables = runnables
+        self._runnable_by_name = {}
+
+        self.max_processes = max_processes
+        self.max_threads = max_threads
+
+    def select_runnables(self, event):
```
Comment (davesh0812) [inline storey/flow.py]: I think typing for the return value can be very helpful here

File: storey/flow.py [hunk 2]
```diff
@@ -1435,3 +1436,182 @@ def get_table(self, key):
...
+    :param runnables: A list of ParallelExecutionRunnable instances.
+    :param max_processes: Maximum number of processes to spawn.
+    :param max_threads: Maximum number of threads to start.
+    """
+
+    def __init__(
+        self,
+        runnables: list[ParallelExecutionRunnable],
+        max_processes: Optional[int] = None,
+        max_threads: Optional[int] = None,
```
Comment (davesh0812) [inline storey/flow.py:1535]: maybe we can add some default

File: storey/flow.py [hunk 3]
```diff
@@ -1435,3 +1436,182 @@ def get_table(self, key):
...
+    Subclasses must also override the run() method with user code that handles the event and returns a result.
+
+    Subclasses may optionally override the init() method if the user's implementation of run() requires prior
+    initialization.
+
+    :param name: Runnable name
+    """
+
+    execution_mechanism = None
+    supported_mechanisms = ("multiprocessing", "threading", "asyncio", "naive")
```
Comment (davesh0812) [inline storey/flow.py]: I think it would be better to move this definition out of the class.

File: storey/flow.py [hunk 4]
```diff
@@ -1435,3 +1436,182 @@ def get_table(self, key):
...
+                        executor,
+                        runnable._run,
+                        input,
+                        event.path,
+                    )
+                futures.append(future)
+            results: list[_ParallelExecutionRunnableResult] = await asyncio.gather(*futures)
+            event.body = {"input": event.body, "results": {}}
+            for result in results:
+                event.body["results"][result.runnable_name] = {"runtime": result.runtime, "output": result.data}
```
Comment (davesh0812) [inline storey/flow.py:1618]: I think we should avoid modifying the original event directly and instead add all relevant information under a new key (self.name represent the name of the current step(ParallelExecution)):
```suggestion
            event.body["result"][self.name] = {"input": input}
            for result in results:
                event.body["results"][self.name][result.runnable_name] = {"runtime": result.runtime, "output": result.data}
```

File: storey/flow.py [hunk 5]
```diff
@@ -1435,3 +1436,182 @@ def get_table(self, key):
...
+            if runnables is None:
+                runnables = self.runnables
+            futures = []
+            runnables_encountered = set()
+            for runnable in runnables:
+                if isinstance(runnable, str):
+                    runnable = self._runnable_by_name[runnable]
+                if id(runnable) in runnables_encountered:
+                    raise ValueError(f"select_runnables() returned more than one outlet named '{runnable.name}'")
+                input = event.body if runnable.execution_mechanism == "multiprocessing" else copy.deepcopy(event.body)
```
Comment (davesh0812) [inline storey/flow.py:1599]: What about the input mechanism, the user can choose some path in the  `event.body`?

**Review by davesh0812 — COMMENTED** [review]
Looks great... few comments :)

File: storey/flow.py (hunk 1 above)
Comment (gtopper) [reply storey/flow.py]: Already done :)

File: storey/flow.py (hunk 2 above)
Comment (gtopper) [reply storey/flow.py:1535]: Not sure how helpful it would be... It could also be confusing if we want to keep supporting `None` as a way to set no limit.

Comment (gtopper) [reply storey/flow.py:1535]: We discussed and decided to add some reasonable limit, but maybe in mlrun and not here (TBD).

File: storey/flow.py (hunk 4 above)
Comment (gtopper) [reply storey/flow.py:1618]: The current output is in line with the planning we've done so far, but we've discussed the fact that we should run this by DS and consider different common scenarios, and maybe need to change this format accordingly.

Therefore, leaving as is in this PR as discussed.

File: storey/flow.py (hunk 5 above)
Comment (gtopper) [reply storey/flow.py:1599]: As discussed, this will be looked at together with the output format.

File: storey/flow.py (hunk 2 above)
Comment (gtopper) [reply storey/flow.py:1535]: Actually, there is already a default in `ProcessPoolExecutor` which is [os.cpu_count() or 1](https://github.com/python/cpython/blob/v3.9.21/Lib/concurrent/futures/process.py#L592), and in `ThreadPoolExecutor` which is [min(32, (os.cpu_count() or 1) + 4)](https://github.com/python/cpython/blob/v3.9.21/Lib/concurrent/futures/thread.py#L142).

Although python 3.13 changes these defaults as per https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor and https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor.

Comment (gtopper) [reply storey/flow.py:1535]: I'm adding explicit defaults in anyway.

File: storey/flow.py [hunk 6]
```diff
@@ -1435,3 +1437,185 @@ def get_table(self, key):
...
+            elif runnable.execution_mechanism == "threading":
+                num_threads += 1
+            elif runnable.execution_mechanism not in ("asyncio", "naive"):
+                raise ValueError(f"Unsupported execution mechanism: {runnable.execution_mechanism}")
+
+        # enforce max
+        if self.max_processes:
+            num_processes = min(num_processes, self.max_processes)
+        if self.max_threads:
+            num_threads = min(num_threads, self.max_threads)
```
Comment (davesh0812) [inline storey/flow.py]: The if statement can be remove - due to the new defaults.  
```suggestion
        num_processes = min(num_processes, self.max_processes)
        num_threads = min(num_threads, self.max_threads)
```

**Review by davesh0812 — APPROVED** [review]
LGTM

File: storey/flow.py [hunk 7]
```diff
@@ -1435,3 +1437,183 @@ def get_table(self, key):
...
+    * "multiprocessing" – To run in a separate process. This is appropriate for CPU or GPU intensive tasks as they
+        would otherwise block the main process by holding Python's Global Interpreter Lock (GIL).
+    * "threading" – To run in a separate thread. This is appropriate for blocking I/O tasks, as they would otherwise
+        block the main event loop thread.
+    * "asyncio" – To run in an asyncio task. This is appropriate for I/O tasks that use asyncio, allowing the event
+        loop to continue running while waiting for a response.
+    * "naive" – To run in the main event loop. This is appropriate only for trivial computation and/or file I/O. It
+        means that the runnable will not actually be run in parallel to anything else.
+
+    Subclasses must also override the run() method with user code that handles the event and returns a result.
```
Comment (assaf758) [inline storey/flow.py]: also need to mention the run vs run_async thing

File: storey/flow.py [hunk 8]
```diff
@@ -1435,3 +1437,183 @@ def get_table(self, key):
...
+
+    def _init(self):
+        super()._init()
+        num_processes = 0
+        num_threads = 0
+        for runnable in self.runnables:
+            if runnable.name in self._runnable_by_name:
+                raise ValueError(f"ParallelExecutionRunnable name '{runnable.name}' is not unique")
+            self._runnable_by_name[runnable.name] = runnable
+            print(f"initializing runnable {runnable.name}")
```
Comment (assaf758) [inline storey/flow.py]: is this print needed?

**Review by assaf758 — COMMENTED** [review]
Looks great!
minor comments

File: storey/flow.py (hunk 8 above)
Comment (gtopper) [reply storey/flow.py]: Not really.

**Review by assaf758 — APPROVED** [review]

**Pull request merged**